src/counts.jl

```diff
@@ -19,21 +19,21 @@ end
     addcounts!(r, x, levels::UnitRange{<:Int}, [wv::AbstractWeights])
 
 Add the number of occurrences in `x` of each value in `levels` to an existing
-array `r`. If a weighting vector `wv` is specified, the sum of weights is used
-rather than the raw counts.
+array `r`.
+
+If a weighting vector `wv` is specified, the sum of weights is used rather than the
+raw counts.
 """
 function addcounts!(r::AbstractArray, x::IntegerArray, levels::IntUnitRange)
-    # add counts of integers from x to r
+    # add counts of integers from x that fall within levels to r
 
-    k = length(levels)
-    length(r) == k || throw(DimensionMismatch())
+    checkbounds(r, axes(levels)...)
 
-    m0 = levels[1]
-    m1 = levels[end]
+    m0 = first(levels)
+    m1 = last(levels)
     b = m0 - 1
 
-    @inbounds for i in 1 : length(x)
-        xi = x[i]
+    @inbounds for xi in x
         if m0 <= xi <= m1
             r[xi - b] += 1
         end
```

Comment:
I'm not sure this actually changes the correctness that much. In case of incorrect indices in `x` this will still segfault or in the wrong case silently corrupt data. Shouldn't `@inbounds` be called only if `x`s are sanitized?

I'd suggest something along those lines:

```julia
Base.@propagate_inbounds function addcounts!(r::AbstractArray, x::IntegerArray, levels::IntUnitRange)
    @boundscheck checkbounds(r, axes(levels)...)
    m0, m1 = first(levels), last(levels)
    b = m0 - 1 # are we sure this is a valid index?
    for xi in x
        if m0 <= xi <= m1
            r[xi - b] += 1
        end
    end
    return r
end
```
where the caller of `addcounts!` takes responsibility for `@inbounds`-ing the call.

Reply:
> incorrect indices in `x`

I'm not sure what you mean here. Could you give an example input that would cause a segfault?

Reply:
`b` is not an index, it's something we subtract from `xi` to get an index. That said, it's a clever observation that `b` is possibly invalid. If `levels` was, for example, a `UnitRange` of `SaferIntegers.UInt`, which throw errors instead of overflowing, then this may be an error. However, we know that `eltype` of levels must be `Int`, and so we can rely on the arithmetic trick the original authors used here to simplify the inner loop.

We could rely on the compiler to factor that subtraction out and manually inline `b`, but I don't see much advantage to including that in this PR.

Reply:
> Shouldn't `@inbounds` be called only if `x`s are sanitized?

Sanitization happens on line 37 `if m0 <= xi <= m1`

Reply:
I think that e.g. `1` may not be a valid index for `r`. Then for `xi=m0` we try to access `r` at `1`?
(and by a segfault I mean invalid read/write to memory)

Reply:
We know from line 37 that `first(levels) <= xi <= last(levels)`.
Subtracting `b = first(levels) - 1` from all sides, we get `1 <= xi-b <= last(levels)-first(levels)+1 = length(levels)`
`levels` is an `IntUnitRange` which is a subtype of `UnitRange`. `UnitRange`s use one-based indexing. Thus `xi-b` is in bounds in `levels`.
We know from line 30 that `axes(levels)` is in bounds in `r`, so anything that is in bounds in `levels` must also be in bounds in `r`.
Therefore `@inbounds r[xi - b]` is safe.

It would be possible to subtract off `firstindex(levels)` and extend support to `AbstractUnitRange{<:Integer}`, but this PR is concerned with correctness, not completeness.

Reply:
ok, so this is correct since we check that indices of `levels` fit into `r`, and the former happens indexes through `OneTo(n)`. TIL ;)